localstack/testing/testselection/matching.py

```diff
@@ -0,0 +1,170 @@
+    for svc, deps in dependency_map.items():
+        for dep in deps:
+            result.setdefault(dep, set()).add(svc)
+    return result
+
+
+def _expand_api_dependencies(svc_name: str) -> set[str]:
+    result = set()
+
+    dependencies = _reverse_dependency_map(API_DEPENDENCIES).get(svc_name, [])
+    result.update(dependencies)
+
+    for dep in dependencies:
+        sub_deps = _expand_api_dependencies(dep)  # recursive call
+        result.update(sub_deps)
+    return result
+
+
+def get_test_dir_for_service(svc: str):
+    return f"tests/aws/services/{svc}"
+
+
+class Matcher:
+    def __init__(self, matching_func: Callable[[str], bool]):
+        self.matching_func = matching_func
+
+    def full_suite(self):
+        return lambda t: [SENTINEL_ALL_TESTS] if self.matching_func(t) else []
+
+    def ignore(self):
+        return lambda t: [SENTINEL_NO_TEST] if self.matching_func(t) else []
+
+    def service_tests(self, services: list[str]):
+        return (
+            lambda t: [get_test_dir_for_service(svc) for svc in services]
+            if self.matching_func(t)
+            else []
+        )
+
+    def passthrough(self):
+        return lambda t: [t] if self.matching_func(t) else []
+
+
+class Matchers:
+    @staticmethod
+    def glob(glob: str) -> Matcher:
+        return Matcher(lambda t: fnmatch.fnmatch(t, glob))
+
+    @staticmethod
+    def regex(glob: str) -> Matcher:
+        return Matcher(lambda t: bool(re.match(t, glob)))
+
+    @staticmethod
+    def prefix(prefix: str) -> Matcher:
+        return Matcher(lambda t: t.startswith(prefix))
+
+
+def generic_service_test_matching_rule(changed_file_path: str) -> set[str]:
+    """
+    Generic matching of changes in service files to their tests
+
+    :param changed_file_path: the file path of the detected change
+    :return: list of partial test file path filters for the matching service and all services it depends on
+    """
+    # TODO: consider API_COMPOSITES
+
+    match = re.findall("localstack/services/([^/]+)/.+", changed_file_path)
+    if not match:
+        match = re.findall(r"localstack/aws/api/([^/]+)/__init__\.py", changed_file_path)
+
+    if match:
+        changed_service = match[0]
+        changed_services = [changed_service]
+        service_dependencies = resolve_dependencies(changed_service)
+        changed_services.extend(service_dependencies)
+        changed_service_module_names = [_map_to_module_name(svc) for svc in changed_services]
+        return {f"tests/aws/services/{svc}/" for svc in changed_service_module_names}
+
+    return set()
+
+
+MatchingRule = Callable[[str], Iterable[str]]
+
+
+def check_rule_has_matches(rule: MatchingRule, files: Iterable[str]) -> bool:
+    """maintenance utility to check if a rule has any matches at all in the given directory"""
+    detected_tests = set()
+    for file in files:
+        detected_tests.update(rule(file))
+    return len(detected_tests) > 0
+
+
+MATCHING_RULES: list[MatchingRule] = [
```

Comment:
We can extract this at least into a separate file if that makes it easier. Other than that I'm having a hard time coming up with a more appropriate location for now besides `localstack.testing.testselection`. :thinking:

Reply:
I guess it makes sense for now. Looking at the next iterations, this could be a super-useful pytest plugin. How would I want to configure that?
Honestly, I am not fully aware of the best practices, but I guess this could be:
- A config file, like an `rc` file somewhere?
- The `conftest` module?
- The `pyproject.toml`?

But again, for now, it's totally fine! 🚀

Reply:
Might indicate that I need to split that file. These matchers are actually not part of the pytest plugin, but of the "preprocessing" that takes the calculated changes and creates the filter file (which is then used by the pytest plugin).

One could rethink that approach and move the mentioned "preprocessing" into the pytest plugin as well and instead of the filter file you provide the file with the list of changes. The drawback would be that we do the calculation of the filters a few times more but it might make the abstraction a bit less confusing in terms of usage.

Which means at least `conftest` with that design for now won't work since it's too late in the process. `rc` file or `pyproject.toml` would be too static for the matcher rule definitions right now since they're code. I'd say we wait until we have stabilized this a bit, maybe we don't even need that flexibility and can encode this in simpler (semi-static) rules.